runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -12229,6 +12229,447 @@ static bool inlineIntrinsicInflate(TR::Node *node, TR::CodeGenerator *cg)
+         TR::MemoryReference::createWithIndexReg(cg, startReg, indexReg, 1));
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, storeReg, storeReg, 0x80);
+   // if the ZERO bit (bit 2) is zero, the value is negative
+   if (p10Plus)
+      {
+      if (isCountPositives) // indexReg = storeReg < 0 ? indexReg : indexReg + 1
+         {
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::setbc, node, tempReg, cr0, 2);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, indexReg, indexReg, tempReg);
+         }
+      else // tempReg = (storeReg < 0)
+         {
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::setbcr, node, tempReg, cr0, 2);
+         }
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, endLabel);
+      }
+   else
+      {
+      // For countPosives, we should return the existing index for a negative value, so we can
+      // jump straight to the end, otherwise we return the length
+      // For hasNegatives, we pre-load a 1 in the case of negatives, and load back a 0 otherwise
+      if (!isCountPositives)
+         generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tempReg, 1);
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, endLabel, cr0);
+      if (isCountPositives)
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, indexReg, indexReg, 1);
+      else
+         generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tempReg, 0);
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, endLabel);
+      }
+
+   // This is actually where we do our main length checks; we don't do it earlier for speed.
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, serialWordCheckLabel);
+   // tempReg marks the end where we can use the vector loop; go there if that's positive.
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addic_r, node, tempReg, lengthReg, -15);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bgt, node, vectorLoopPrepLabel, cr0);
+
+   // check if we can load double words, but first do some common code
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, tempReg, indexReg, lengthReg);
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, serialDWordCheckLabel);
+   // we want to load 0x80808080 into maskReg, but lis was designed for signed values,
+   // and would throw an error for 0x8080, yet it could accept the equivalent negative value of it;
+   // we don't worry about sign extension since the upper word should be 0 in storeReg after lwzx
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::lis, node, maskReg, -32640);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ori, node, maskReg, maskReg, 0x8080);
+
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, cr0, tempReg, 8);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bge, node, serialDWordLabel, cr0);
```

Comment:
i think the sequence can be laid down better to reduce/minimize the number of taken branch (each taken branch is worth of roughly 3-cycles.  at this point, it is known that there is at most one dw, word, half, and byte respectively. jump to dw here, you need to jump back to handle 0-7 residue later anyway? let dw fall-through here appears better